[DOC] Add join processor documentation (#5985)
opensearch-project/documentation-website: _ingest-pipelines/processors/join.md

---
layout: default
title: Join
parent: Ingest processors
nav_order: 160
---

# Join processor

> The `join` processor combines fields from different data sources into a single document before indexing. Raises an error if the field is not an array. For example, you could combine the log message, severity level, and timestamp into a single field for better readability and easier querying. Or, you could join data from different sources, such as application logs and system logs, based on a common field like a session ID or user ID to provide a more comprehensive view of related events and help in troubleshooting and root cause analysis. 

Reviewer: `join` processor can only be used to join every elements in an array into a single string value, join data from different sources seems not possible, see the [comment](https://github.com/opensearch-project/OpenSearch/blob/27003270adedebd43eedbd0186c0825f2c0e8613/modules/ingest-common/src/main/java/org/opensearch/ingest/common/JoinProcessor.java#L45) in the code.

Author: Revised.

The following is the syntax for the `join` processor:

```json
{
  "join": {
    "field": "field_name",
    "separator": "separator_string"
  }
}
```
{% include copy-curl.html %}

## Configuration parameters

The following table lists the required and optional parameters for the `join` processor.

Parameter | Required/Optional | Description |
|-----------|-----------|-----------|

> `field` | Required | The field name where the join operator is applied. 

Reviewer: `field` must be an array, we should emphasize this.

Author: Revised

> `separator` | Optional | A string separator to use when joining field values. If not specified, the values are concatenated without a separator.

Reviewer: `separator` is required.

`target_field` | Optional | The field to assign the cleaned value to. If not specified, field is updated in-place.
`description` | Optional | Description of the processor's purpose or configuration.
`if` | Optional | Conditionally execute the processor.
`ignore_failure` | Optional | Ignore failures for the processor. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).
`on_failure` | Optional | Handle failures for the processor. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).
`tag` | Optional | Identifier for the processor. Useful for debugging and metrics.

## Using the processor

Follow these steps to use the processor in a pipeline.

### Step 1: Create a pipeline

The following query creates a pipeline, named <name>, that uses the `join` processor to <do what?>: 

```json

> PUT _ingest/pipeline/example-join-pipeline

Reviewer: ```
PUT _ingest/pipeline/example-join-pipeline
{
  "description": "Example pipeline using the join processor",
  "processors": [
    {
      "join": {
        "field": "uri",
        "separator": "/"
      }
    }
  ]
}
```

{
  "description": "Example pipeline using the join processor",
  "processors": [
    {
      "join": {
        "field": "message",
        "separator": " - ",
        "target_field": "combined_message"
      }
    }
  ]
}
```
{% include copy-curl.html %}

### Step 2 (Optional): Test the pipeline

It is recommended that you test your pipeline before you ingest documents.
{: .tip}

To test the pipeline, run the following query:

```json

> POST _ingest/pipeline/example-join-pipeline/_simulate

Reviewer: ```
POST _ingest/pipeline/example-join-pipeline/_simulate
{
  "docs": [
    {
      "_source": {
        "uri": [
          "app",
          "home",
          "overview"
        ]
      }
    }
  ]
}
```

{
  "docs": [
    {
      "_source": {
        "message": "Server started",
        "severity": "INFO",
        "timestamp": "2023-04-20T12:00:00Z"
      }
    }
  ]
}
```
{% include copy-curl.html %}

#### Response

The following example response confirms that the pipeline is working as expected:

```json

> 

Reviewer: ```
{
  "docs": [
    {
      "doc": {
        "_index": "_index",
        "_id": "_id",
        "_source": {
          "uri": "app/home/overview"
        },
        "_ingest": {
          "timestamp": "2024-05-24T02:16:01.00659117Z"
        }
      }
    }
  ]
}
```

```

### Step 3: Ingest a document 

The following query ingests a document into an index named `testindex1`:

```json
<insert code example>
```

> {% include copy-curl.html %}

Reviewer: ```
POST testindex1/_doc/1?pipeline=example-join-pipeline
{
  "uri": [
    "app",
    "home",
    "overview"
  ]
}
```

> #### Response

Reviewer: Seems this section is not needed.

The request indexes the document into the index <index name> and will index all documents with <what does this response tell the user?>.

```json
<insert code example>
```

### Step 4 (Optional): Retrieve the document

To retrieve the document, run the following query:

```json

> <insert code example>

Reviewer: `GET testindex1/_doc/1`

---
layout: default
title: Join
parent: Ingest processors
nav_order: 160
---

# Join processor

The `join` processor concatenates the elements of an array into a single string value, using a specified separator between each element. It throws an exception if the provided input is not an array.

The following is the syntax for the `join` processor:

```json
{
  "join": {
    "field": "field_name",
    "separator": "separator_string"
  }
}
```
{% include copy-curl.html %}

## Configuration parameters

The following table lists the required and optional parameters for the `join` processor.

Parameter | Required/Optional | Description |
|-----------|-----------|-----------|

> `field` | Required | The field name where the join operator is applied. Must be an array.

Reviewer 2: ```suggestion
`field` | Required | The name of the field to which the join operator is applied. Must be an array.
```

> `separator` | Required | A string separator to use when joining field values. If not specified, the values are concatenated without a separator.

Reviewer 2: ```suggestion
`separator` | Required | A string separator to use when joining field values. If not specified, then the values are concatenated without a separator.
```

> `target_field` | Optional | The field to assign the cleaned value to. If not specified, field is updated in-place.

Reviewer 2: ```suggestion
`target_field` | Optional | The field to assign the cleaned value to. If not specified, then the field is updated in place.
```

> `description` | Optional | Description of the processor's purpose or configuration.

Reviewer 2: ```suggestion
`description` | Optional | A description of the processor's purpose or configuration.
```

> `if` | Optional | Conditionally execute the processor.

Reviewer 2: ```suggestion
`if` | Optional | Specifies to conditionally execute the processor.
```

> `ignore_failure` | Optional | Ignore failures for the processor. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).

Reviewer 2: ```suggestion
`ignore_failure` | Optional | Specifies to ignore failures for the processor. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).
```

> `on_failure` | Optional | Handle failures for the processor. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).

Reviewer 2: ```suggestion
`on_failure` | Optional | Specifies to handle failures for the processor. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).
```

> `tag` | Optional | Identifier for the processor. Useful for debugging and metrics.

Reviewer 2: ```suggestion
`tag` | Optional | An identifier for the processor. Useful for debugging and metrics.
```

## Using the processor

Follow these steps to use the processor in a pipeline.

### Step 1: Create a pipeline

> The following query creates a pipeline, named <name>, that uses the `join` processor to <do what?>: 

Author: ```suggestion
The following query creates a pipeline, named <name>, that uses the `join` processor to <do what?>: 
```
```suggestion
The following query creates a pipeline named `example-join-pipeline` that uses the `join` processor to concatenate all the values of the `uri`  field, separating them with the specified separator `/`: 
```

---
layout: default
title: Join
parent: Ingest processors
nav_order: 160
---

# Join processor

The `join` processor is used to <explain what is used to do>.

The following is the syntax for the `join` processor:

```json
<insert syntax example>
```
{% include copy-curl.html %}

## Configuration parameters

The following table lists the required and optional parameters for the `join` processor.

Parameter | Required/Optional | Description |
|-----------|-----------|-----------|
<insert the parameters>

## Using the processor

Follow these steps to use the processor in a pipeline.

### Step 1: Create a pipeline

The following query creates a pipeline, named <name>, that uses the `join` processor to <do what?>: 

```json
<insert pipeline code example>
```
{% include copy-curl.html %}

### Step 2 (Optional): Test the pipeline

It is recommended that you test your pipeline before you ingest documents.
{: .tip}

To test the pipeline, run the following query:

```json
<insert code example>

> ```

Author: @gaobinlong Could you help me troubleshoot this pipleline example? I'm getting the following error: {
  "docs": [
    {
      "error": {
        "root_cause": [
          {
            "type": "illegal_argument_exception",
            "reason": "field [message] of type [java.lang.String] cannot be cast to [java.util.List]"
          }
        ],
        "type": "illegal_argument_exception",
        "reason": "field [message] of type [java.lang.String] cannot be cast to [java.util.List]"
      }
    }
  ]
}
